jax/interpreters/partial_eval.py

```diff
@@ -462,10 +468,17 @@ def instantiate_const_at(trace, instantiate: bool, tracer):
 FreeVar = namedtuple('FreeVar', ['val'])
 ConstVar = namedtuple('ConstVar', ['val'])
 LambdaBinding = namedtuple('LambdaBinding', [])
-JaxprEqnRecipe = namedtuple('JaxprEqnRecipe',
-                            ['eqn_id', 'invars', 'outvars', 'primitive', 'params'])
-
-def new_eqn_recipe(invars, outvars, primitive, params):
+class JaxprEqnRecipe(NamedTuple):
+  eqn_id: object
+  invars: Sequence[JaxprTracer]
+  outvars: 'Sequence[ref[JaxprTracer]]'
+  primitive: core.Primitive
+  params: Dict[str, Any]
+
+def new_eqn_recipe(invars: Iterable[JaxprTracer],
```

Comment:
We've been using `Sequence` for parameter types. `Iterable` seems to be more general, should we standardize on one?

Reply:
I think `Sequence` is better because iterables are mutation-y and bad in general IMO. That is, Python 3's `map` produces an iterable, as does a generator expression, but iterating over those things mutate their state! I believe `Sequence` means "an iterable that doesn't mutate itself during iteration."

Reply:
At least one caller passes an iterator, not a sequence.
https://github.com/google/jax/blob/49a441f7456b4090ac1cd74668d9638da0aca59b/jax/interpreters/partial_eval.py#L295

I thought the caller wasn't wrong to pass an iterator here. But perhaps we should simply say `tuple`, given we are going to convert the result into a tuple anyway?

Reply:
Yeah I guess it works to pass an iterator, but I think it's unsafe and it might be getter to replace that with `(*const_tracers, *env_tracers)` or similar.

Reply:
I updated the caller to form a tuple and changed this to `Sequence`.